### PR TITLE
Custom Build Outputs `--output`

### LINKED ISSUE #415: Custom Build Outputs `--output`
I would like to directly output the Docker build to local disk without creating an intermediate image. For this, Docker/BuildKit has the option of specifying an `--output` flag ([docs](https://docs.docker.com/reference/cli/docker/image/build/#output)). I tried to hunt around bollard, but it seems to me that this API isn't exposed at the moment. Since the Docker Engine API version bollard is targeting already [includes this feature](https://docs.docker.com/engine/api/v1.44/#tag/Image/operation/ImageBuild), I was wondering whether this is already possible and I simply didn't find it in the docs or else what's required to make this work. Thanks :)

### PR DESCRIPTION
**Pull request by levinwinter**
Adds the ability to specify the exporter (currently `tar` and `local`). Details and a more thorough description is in the issue. Let me know what you think! I'm not so sure about `codegen/proto/src/lib.rs`... Feedback is welcome :)

Edit: I'm not sure how to deal with existing files/directories. If you have suggestions I can explicitly handle that case!

Closes #415

### CODE REVIEW

**Comment by levinwinter** [comment]
I'm quite sure that some edge-cases are not handled well for the local exporter (i.e. implementation of `fsutil`). I think it's good enough for what I need, but let me know what sort of maturity you expect.

**Comment by fussybeaver** [comment]
Fab! This must have been quite a headache to get to work. I will need to take some time to test your branch locally.

I am a little bit surprised that we've needed to circumvent filesend protobuf generated code. I'm wondering whether that's a lacking functionality in tonic or something that buildkit is doing.. did you find any issues / tickets in other repos about this?

File: codegen/proto/src/lib.rs [hunk 1]
```diff
@@ -40,6 +40,179 @@ pub mod moby {
...
+                    _cx: &mut Context<'_>,
+                ) -> Poll<std::result::Result<(), Self::Error>> {
+                    Poll::Ready(Ok(()))
+                }
+                fn call(&mut self, req: http::Request<B>) -> Self::Future {
+                    let _inner = self.inner.clone();
+                    match req.uri().path() {
+                        "/moby.filesync.v1.FileSend/DiffCopy" => {
+                            #[allow(non_camel_case_types)]
+                            struct DiffCopySvc<T: FileSendPacket>(pub Arc<T>);
```
Comment (fussybeaver) [inline codegen/proto/src/lib.rs:137]: I see the inner type is different here than the generated `FileSend` type. I guess that's what you need to send back to buildkit ?

Comment (levinwinter) [reply codegen/proto/src/lib.rs:137]: Yes, `crate::fsutil::types::Packet` is the correct message type for the gRPC stream! We both send and receive these packets when using the `local` exporter (as opposed to `BytesMessage` for the `tar` exporter).

**Comment by levinwinter** [comment]
You are absolutely right, it was quite a headache :P

I did some research and could find this issue: https://github.com/moby/buildkit/issues/2109. What they describe there matches my understanding! `fsutil` can use a "raw" gRPC stream and it seems to care little what is written in the protobuf definition. Unfortunately, both versions of the service (i.e. sending/receiving `Packet` or `BytesMessage`) would be required to support both the `local` and `tar` exporter, respectively. So we can't just change the definition to use `fsutil.types.Packet` like in the issue's PR.

And as for using "raw" gRPC streams in tonic, I couldn't find anything online. I was also thinking whether the existing generated code could be reused in some way (since the implementations are almost the same, except a few lines), but I also didn't manage to do that.

File: codegen/proto/src/lib.rs [hunk 2]
```diff
@@ -40,6 +40,179 @@ pub mod moby {
     pub mod filesync {
         pub mod v1 {
             include!("generated/moby.filesync.v1.rs");
+
```
Comment (fussybeaver) [inline codegen/proto/src/lib.rs:43]: Could we move this part of the code into another file and `include!` it here - simply makes it easier for me maintaining this file as it sometimes gets new generated files.

**Comment by fussybeaver** [comment]
This looks good, and happy to merge it after addressing a small cleanup. 

We will need to add some additional `local` / `tar` export types for the "driver"-based API similar to what we use for the OCI exporter, but happy to address that down the road in a subsequent PR...

File: codegen/proto/src/lib.rs (hunk 2 above)
Comment (levinwinter) [reply codegen/proto/src/lib.rs]: I changed things around a bit and everything is now incorporated into the `codegen/proto` infrastructure! That documents the process in a way and also gives the new code a home (which I added using `include!`).

**Comment by levinwinter** [comment]
Sounds good! I refactored the `include!` things. If you think it's okay, feel free to merge :)

**Pull request merged**